src/Compilers/Core/Portable/Operations/OperationExtensions.cs

```diff
@@ -36,8 +36,15 @@ internal static bool HasErrors(this IOperation operation, Compilation compilatio
             }
 
             // if wrong compilation is given, GetSemanticModel will throw due to tree not belong to the given compilation.
-            var model = operation.SemanticModel ?? compilation.GetSemanticModel(operation.Syntax.SyntaxTree);
-            if (model.IsSpeculativeSemanticModel)
+            var model = operation.SemanticModel;
+
+            // An IOperation tree for a simple program includes statements from all compilation units involved,
+            // but each model is tied to a single syntax tree.
+            if (model is null || model.SyntaxTree != operation.Syntax.SyntaxTree)
+            {
+                model = compilation.GetSemanticModel(operation.Syntax.SyntaxTree);
+            }
+            else if (model.IsSpeculativeSemanticModel)
```

Comment:
>if (model.IsSpeculativeSemanticModel) [](start = 17, length = 37)

Is this never true for top-level statements?

Reply:
> Is this never true for top-level statements?

There is no way to speculate on a an entire Simple Program. One can speculate on a statement, but that wouldn't be based on multiple syntax trees.
Nerveless, I agree it would be more robust long term to have this as separate ```if``` statement rather than an ```else if```. I'll make the change, at least in the next PR.

---
In reply to: [384691246](https://github.com/dotnet/roslyn/pull/41926#discussion_r384691246) [](ancestors = 384691246)